js/src/common/Model.ts

```diff
@@ -0,0 +1,383 @@
+import app from '../common/app';
+import { FlarumRequestOptions } from './Application';
+import Store, { ApiPayloadSingle, ApiResponseSingle } from './Store';
+
+interface ModelIdentifier {
+  type: string;
+  id: string;
+}
+
+interface ModelAttributes {
+  [key: string]: unknown;
+}
+
+interface ModelRelationships {
+  [relationship: string]: {
+    data: ModelIdentifier | ModelIdentifier[];
+  };
+}
+
+export interface UnsavedModelData {
+  type?: string;
+  attributes?: ModelAttributes;
+  relationships?: ModelRelationships;
+}
+
+export interface SavedModelData {
+  type: string;
+  id: string;
+  attributes?: ModelAttributes;
+  relationships?: ModelRelationships;
+}
+
+export type ModelData = UnsavedModelData | SavedModelData;
+
+interface SaveRelationships {
+  [relationship: string]: Model | Model[];
+}
+
+interface SaveAttributes {
+  [key: string]: unknown;
+  relationships?: SaveRelationships;
+}
+
+/**
+ * The `Model` class represents a local data resource. It provides methods to
+ * persist changes via the API.
+ */
+export default abstract class Model {
+  /**
+   * The resource object from the API.
+   */
+  data: ModelData = {};
+
+  /**
+   * The time at which the model's data was last updated. Watching the value
+   * of this property is a fast way to retain/cache a subtree if data hasn't
+   * changed.
+   */
+  freshness: Date = new Date();
+
+  /**
+   * Whether or not the resource exists on the server.
+   */
+  exists: boolean = false;
+
+  /**
+   * The data store that this resource should be persisted to.
+   */
+  protected store: Store | null;
+
+  /**
+   * @param data A resource object from the API.
+   * @param store The data store that this model should be persisted to.
+   * @public
+   */
+  constructor(data: ModelData = {}, store = null) {
+    this.data = data;
+    this.store = store;
+  }
+
+  /**
+   * Get the model's ID.
+   *
+   * @final
+   */
+  id(): string | undefined {
+    return 'id' in this.data ? this.data.id : undefined;
+  }
+
+  /**
+   * Get one of the model's attributes.
+   *
+   * @final
+   */
+  attribute<T = unknown>(attribute: string): T {
+    return this.data?.attributes?.[attribute] as T;
+  }
+
+  /**
+   * Merge new data into this model locally.
+   *
+   * @param data A resource object to merge into this model
+   */
+  pushData(data: ModelData | { relationships?: SaveRelationships }): this {
```

Comment:
I did a lot of refactoring on this method to make it more understandable, less hacky, and more typesafe. This one needs a bit of extra review.